### ISSUE TITLE
macOS build?

### ISSUE THREAD

**Issue by DaniruKun** [ButzYung/SystemAnimatorOnline#20]
It is an Electron app, there is 0 reason for there not to be a macOS build, especially when you include Linux builds, which is a less widespread desktop OS.

**Comment by ButzYung** [comment]
I would love to make an macOS build, as you are right, porting electron to macOS should be quite an easy task. But the problem is... I don't have a Mac machine to test, lol

**Comment by DaniruKun** [comment]
@ButzYung I can give it a shot, and I think Github also supports macOS runners now?

**Comment by ButzYung** [comment]
Yeah this app's setup is pretty barebone and non-standard. If you want to try making a macOS build for testing, do the following.

1. Download either Windows or Linux build of XR Animator. Unzip it.
2. [Download an official electron build for macOS.](https://github.com/electron/electron/releases) v22 is recommended as this is what XR Animator is currently using. Unzip it inside the same folder of XR Animator.
3. On the `resources` folder inside the default `electron` folder of XR Animator, copy everything except `default_app.asar` to the same folder under the macOS `electron`.

Finally, run macOS `electron`. If we are lucky enough, you will see the START button of XR Animator as usual without errors. After that, do everything as usual. There may be issues with the tray conext menu, the legacy settings panel, and possibly the saving of app settings, but beyond that, most features should work like normal.

**Comment by DaniruKun** [comment]
Which Electron of a release do I download? There are quite a few versions for the same aarch:
![image](https://github.com/ButzYung/SystemAnimatorOnline/assets/5202322/f8c1bf05-510c-4841-9ffa-d1ef66f69c2e)

Also, I assume the Windows/Linux builds are on x64? This might be a problem for Apple Silicon Macs, so really a long run solution is needed here...

**Comment by ButzYung** [comment]
Should be `electron-v22.3.18-darwin-arm64.zip` I suppose. Yeah Windows/Linux builds are on x64, but that's mainly for my own testing convenience, and XR Animator itself doesn't really have any native codes that rely on the x64 architecture.

**Comment by Vanyanie** [comment]
Any news about running it on macOS? Would like to use it on mac too

**Comment by Vanyanie** [comment]
ok, got it to run on macOS but you need to request for permission to use the camera in eleron on macOS.

https://stackoverflow.com/questions/72024011/electron-app-not-asking-for-camera-and-microphone-permission-on-macos-monterey

So, you need to add this line in your app: 

`systemPreferences.askForMediaAccess("camera")`

https://www.electronjs.org/docs/latest/api/system-preferences#systempreferencesaskformediaaccessmediatype-macos

But I'm not familiar with Elecron or how to code for it or hoe to build the app with it.

**Comment by ButzYung** [comment]
@Vanyanie Wowww really thanks for the testing. As expected XR Animator should work on macOS with very little changes, but what hinders me is that I don't have a Mac to test it. Feel free to test all other features of XR Animator, and see if they work.

According to electron docs, the tray menu of XR Animator should be accessible somewhere on the top right. You can check DevTools there and see if there is any error message there.

**Comment by ThatKoza** [comment]
> [@Vanyanie](https://github.com/Vanyanie) Wowww really thanks for the testing. As expected XR Animator should work on macOS with very little changes, but what hinders me is that I don't have a Mac to test it. Feel free to test all other features of XR Animator, and see if they work.
> 
> According to electron docs, the tray menu of XR Animator should be accessible somewhere on the top right. You can check DevTools there and see if there is any error message.

if you need mac testers, im open to test it for you on intel mac

**Comment by ButzYung** [comment]
@ThatKoza First of all, thanks for your interest to help. Do you have some basic knowledge of programming (JavaScript or something silimar), or any experience in dealing with basic scripting and front-end web development (probably copy-n-paste level will do)?

To be honest, I am looking for someone who can maintain at least a certain level of steady communication during the testing, like I can expect a response with a few hours or 2-3 days at most, instead of months or simply never (like the case right above you).

If you think you can do it, please let me know.

**Comment by ThatKoza** [comment]
> @ThatKoza First of all, thanks for your interest to help. Do you have some basic knowledge of programming (JavaScript or something silimar), or any experience in dealing with basic scripting and front-end web development (probably copy-n-paste level will do)?
> 
> To be honest, I am looking for someone who can maintain at least a certain level of steady communication during the testing, like I can expect a response within a few hours or 2-3 days at most, instead of months or simply never (like the case right above you).
> 
> If you think you can do it, please let me know.

I sadly do not have good knowledge in coding sphere, but i had experience helping a person to get macos support via reporting log file and behaviour. If yoj explain what i need to do, my lack of knowledge wont be a problem, just want to help in any way.

**Comment by ButzYung** [comment]
@ThatKoza I think you have the skills required. No coding knowledge or experience is fine indeed. Just expect some replacement of files here and there, simple editing of scripts with a text editor, and maybe reporting what you see in the console log.

Maybe you can try the methods I described in the earlier posts here, but instead of looking for `v22` of electron, find `v34.1.1` or below and pick the macOS build that works on your system. First, just run `electron` and confirm that it can run (meaning this is the correct build). Then you can put it to the XR Animator folder as instructed, and if you are lucky enough, you may be able to get it starting like the screenshot above. We can begin from there.
https://github.com/electron/electron/releases

**Comment by ThatKoza** [comment]
> @ThatKoza I think you have the skills required. No coding knowledge or experience is fine indeed. Just expect some replacement of files here and there, simple editing of scripts with a text editor, and maybe reporting what you see in the console log.
> 
> Maybe you can try the methods I described in the earlier posts here, but instead of looking for `v22` of electron, find `v34.1.1` or below and pick the macOS build that works on your system. First, just run `electron` and confirm that it can run (meaning this is the correct build). Then you can put it to the XR Animator folder as instructed, and if you are lucky enough, you may be able to get it starting like the screenshot above. We can begin from there.
> https://github.com/electron/electron/releases

Okay, i will start doing that

**Comment by ThatKoza** [comment]
> @ThatKoza I think you have the skills required. No coding knowledge or experience is fine indeed. Just expect some replacement of files here and there, simple editing of scripts with a text editor, and maybe reporting what you see in the console log.
> 
> Maybe you can try the methods I described in the earlier posts here, but instead of looking for `v22` of electron, find `v34.1.1` or below and pick the macOS build that works on your system. First, just run `electron` and confirm that it can run (meaning this is the correct build). Then you can put it to the XR Animator folder as instructed, and if you are lucky enough, you may be able to get it starting like the screenshot above. We can begin from there.
> https://github.com/electron/electron/releases

Okay, so, i got the electron-v34.1.1-mas-x64.zip (since im on intel), moved it into the XR-Animator_v0.27.4b_windows along with licenses and readme. It asks to execute Users/fixed/Downloads/XR-Animator_v0.27.4b_windows/Electron.app/Contents/MacOS/Electron path-to-app, i go to package contents of the electron app, execute it manually and it returns with terminal ![image](https://github.com/user-attachments/assets/3a6fa146-714b-44c4-99dc-8557dcb85eb4)

**Comment by ButzYung** [comment]
Did you copy the files like descibed below?

>On the `resources` folder inside the default electron folder of XR Animator, copy everything except `default_app.asar `to the same folder under the macOS electron.

How about if you run electron raw at its default state? Does the default window show up?

**Comment by ThatKoza** [comment]
> Did you copy the files like descibed below?
> 
> >On the `resources` folder inside the default electron folder of XR Animator, copy everything except `default_app.asar `to the same folder under the macOS electron.
> 
> How about if you run electron raw at its default state? Does the default window show up?
> 
> 
> 
> 

Didn't copy contents of the xr animator resources intially. In raw state it still was asking to execute same path name. After i moved 2 icons with app.asar into /Users/fixed/Downloads/XR-Animator_v0.27.4b_windows/Electron.app/Contents/Resources. Then i run the electron app, it gives an error, the same happens if running electron as executable that it was asking to execute earlier. ![image](https://github.com/user-attachments/assets/89b019bf-8e2d-4c5d-88e1-70144b2af6e9)

**Comment by ButzYung** [comment]
So at least you can get the alert window. That's some progress lol. It seems the MacOS electron has a different folder structure from the Windows/Linux version. I will need to update `app.asar` file for that. I will send you the updated file later for testing.

**Comment by ThatKoza** [comment]
> So at least you can get the alert window. That's some progress lol. It seems the MacOS electron has a different folder structure from the Windows/Linux version. I will need to update `app.asar` file for that. I will send you the updated file later for testing.

Take your time. Overall, the folder system is only different for .app's

**Comment by ButzYung** [comment]
> Take your time. Overall, the folder system is only different for .app's

I wonder if you move/copy `Electron` file from `MacOS` folder one level back to `Contents`, the app still runs or not. That would save some heaches of changing `app.asar` file if it works.

**Comment by ThatKoza** [comment]
> > Take your time. Overall, the folder system is only different for .app's
> 
> I wonder if you move/copy `Electron` file from `MacOS` folder one level back to `Contents`, the app still runs or not. That would save some heaches of changing `app.asar` file if it works.
> 
> 

Nope, it refuses to even give an error pop out  and says that it cannot find the path. Same thing if you run it inside the contents or outside it

**Comment by ButzYung** [comment]
@ThatKoza Unzip the new `app.asar`, replace the old one, and try again.

[app.zip](https://github.com/user-attachments/files/18867609/app.zip)

**Comment by ThatKoza** [comment]
path name is now shorter, lol, but still gives an error <img width="262" alt="Image" src="https://github.com/user-attachments/assets/747af899-dfad-430d-a0b2-6116ee4f3baf" />

**Comment by ButzYung** [comment]
@ThatKoza Try this.

[app.zip](https://github.com/user-attachments/files/18868102/app.zip)

**Comment by ThatKoza** [comment]
it works now, it looks a little funny <img width="1182" alt="Image" src="https://github.com/user-attachments/assets/c0ea3624-252b-4d55-a06a-67efb7e8c3c6" />
transparency is kinda working, but the background is with red tint (personally, i prefer having no transparency at all). 

<img width="963" alt="Image" src="https://github.com/user-attachments/assets/86a393e5-7ff3-4651-bf1b-96896d048a40" />
camera is getting reconised, even the iphone camera is getting found, but im not sure that it is giving any output. the light on the webcam is not getting lit either

<img width="350" alt="Image" src="https://github.com/user-attachments/assets/cf436f7c-93c3-4cb5-94bc-815d3cadf096" />

**Comment by ThatKoza** [comment]
> [@ThatKoza](https://github.com/ThatKoza) That's a lot of progress already. At least the basics work. Webcam not working is expected. As mentioned by other users in the first few posts, some permission settings need to be changed.
> 
> For now, there are things I want to know if it works or not.
> 
> 1. Settings save/load - Try drop your own model at app starts, change some settings, etc, restart the app, and see if settings are properly saved and loaded.
> 2. Tray menu - Try to locate the tray menu of XRA on your MacOS UI, and if it is there, click `DevTools` and see if it works. `DevTools` will be an important tool for future debugging. You may also try basic functions "Restart" and see if it works.
> 3. Try mocap. Since webcam doesn't work for now, try local media file and see how it performs.

1.Settings, model and anything else doesnt save, same problem if i run the windows version via crossover/wine.
2. DevTools works
3. Mocap isnt working, at least with the file that is in xr animator in folder of the js

**Comment by ThatKoza** [comment]
camera works!!!!!!!!!!!!! but, setting up custom wallpaper doesnt save it, even if clicking apply. (tried it with different exstensions). 
Should i do showcase even with red background or let you do a fix for wallpapers?

**Comment by ButzYung** [comment]
@ThatKoza Glad to know webcam works. Or maybe under certain system/conditions? Afterall, the info up there was more than a year old. Maybe something has changed.

So I suppose settings save/load work now?

For now, go to `UI/Options➡️Scene➡️Background image`, drop an image and it will become the app's wallpaper. While it's ok to touch those settings on legacy settings panel for testing, for practical usage I won't recommend it. Most of those settings are obsolete (even worse, some of those settings are Windows specific) and don't have much to with XR Animator anyways.

**Comment by ThatKoza** [comment]
> @ThatKoza Glad to know webcam works. Or maybe under certain system/conditions? Afterall, the info up there was more than a year old. Maybe something has changed.
> 
> So I suppose settings save/load work now?
> 
> For now, go to `UI/Options➡️Scene➡️Background image`, drop an image and it will become the app's wallpaper. While it's ok to touch those settings on legacy settings panel for testing, for practical usage I won't recommend it. Most of those settings are obsolete (even worse, some of those settings are Windows specific) and don't have much to do with XR Animator anyways.

I haven't tested input from iPhone camera yet, but built in camera works, yes. 
Heres kinda bootleggy footage: https://drive.google.com/drive/folders/1v1qFVezq5YeZ41GZ4J7fuqgMPZnMNO0v (if ill be honest, it would be better to do motion on your end, my lighting is shit)

**Comment by ButzYung** [comment]
@ThatKoza Can you capture video that includes the desktop as well (partial dekstop is fine if you don't want to show full desktop, as long as people can tell it's a Mac)? I think any motion is fine lol, or maybe you can use a dance video from YouTube or something.

I wonder what's the spec of your Mac PC. The frame rate looks lower than expected.

**Comment by ButzYung** [comment]
Are you using OBS for screen capture? Nowadays screen capturing software shouldn't affect frame rate much.

I am curious about the GPU info of your system. Open the legacy settings panel, `Mode` tab, click `GPU check`, and you can download the log there.

**Comment by ThatKoza** [comment]
> Are you using OBS for screen capture? Nowadays screen capturing software shouldn't affect frame rate much.
> 
> I am curious about the GPU info of your system. Open the legacy settings panel, `Mode` tab, click `GPU check`, and you can download the log there.

affects mine though, lol, cant do anything about that

[about-gpu-2025-02-19T17-40-15-263Z.txt](https://github.com/user-attachments/files/18873064/about-gpu-2025-02-19T17-40-15-263Z.txt)

**Comment by ButzYung** [comment]
Or maybe you can include a still shot of your desktop as the app's wallpaper, whatever way that makes it obvious it's from a Mac.

**Comment by ThatKoza** [comment]
![image](https://github.com/user-attachments/assets/d6c5e4c6-60ee-4272-8389-479db22872f1)![image](https://github.com/user-attachments/assets/7d38374d-69a4-45d2-a770-cf57826c03a6)![image](https://github.com/user-attachments/assets/38489b99-8c45-4e55-9879-0afd72a8c525)![image](https://github.com/user-attachments/assets/7f33c56c-c1eb-4d8a-96b9-c4e67339978a)
Is anything else needed in the screenshot?

**Comment by ButzYung** [comment]
@ThatKoza Do you know your macOS version. According to the following page, special setup to request user consent for webcam access is not required until macOS 10.14 Mojave. This may explain why your webcam is accessible without modification to the app.
https://www.electronjs.org/docs/latest/api/system-preferences#systempreferencesaskformediaaccessmediatype-macos

**Comment by ThatKoza** [comment]
> @ThatKoza Do you know your macOS version? According to the following page, special setup to request user consent for webcam access is not required until macOS 10.14 Mojave. This may explain why your webcam is accessible without modification to the app.
> https://www.electronjs.org/docs/latest/api/system-preferences#systempreferencesaskformediaaccessmediatype-macos

Sequoia 15.2

**Comment by ButzYung** [comment]
So maybe it's because of something else. I will need more beta testers to confirm if changes are required.

**Comment by ThatKoza** [comment]
> So maybe it's because of something else. I will need more beta testers to confirm if changes are required.

Changes would probably be required if they are running on MacOs version lower than Mojave. I myself cannot test that, since my older MacBook is running on 12.7.4 Monterey (which by itself is already pretty old)

**Comment by ButzYung** [comment]
@ThatKoza Please unzip the file, and just like the previous case, copy the JS file to `AT_SystemAnimator_vxxxx.gadget\js`, overwritting existing files. 

[SA_system_emulation.min.zip](https://github.com/user-attachments/files/18883675/SA_system_emulation.min.zip)

This should solve the webcam permission issue for all macOS version. It probably may not do anything in your case since you may have the permission already, but I want to confirm that at least it continues to work for you. Try mocap, access your camera as usual and see if there is any error.

**Comment by ThatKoza** [comment]
> [@ThatKoza](https://github.com/ThatKoza) Please unzip the file, and just like the previous case, copy the JS file to `AT_SystemAnimator_vxxxx.gadget\js`, overwritting existing files.
> 
> [SA_system_emulation.min.zip](https://github.com/user-attachments/files/18883675/SA_system_emulation.min.zip)
> 
> This should solve the webcam permission issue for all macOS version. It probably may not do anything in your case since you may have the permission already, but I want to confirm that at least it continues to work for you. Try mocap, access your camera as usual and see if there is any error.

in fact, it did something for me. camera isnt getting recognised at all now, lol. you propably should make a seperate release for macos lower than mojave, because it just looks like it breaks in newer versions.

**Comment by ButzYung** [comment]
> in fact, it did something for me. camera isnt getting recognised at all now, lol. you propably should make a seperate release for macos lower than mojave, because it just looks like it breaks in newer versions.

SORRY I made a typo in the script LOL. This one should work.

[SA_system_emulation.min.zip](https://github.com/user-attachments/files/18883824/SA_system_emulation.min.zip)

**Comment by ThatKoza** [comment]
> > in fact, it did something for me. camera isnt getting recognised at all now, lol. you propably should make a seperate release for macos lower than mojave, because it just looks like it breaks in newer versions.
> 
> SORRY I made a typo in the script LOL. This one should work.
> 
> [SA_system_emulation.min.zip](https://github.com/user-attachments/files/18883824/SA_system_emulation.min.zip)

ok, now it works, ahahahahah. then i doubt that it needs a separate release (unless someone with old macos will be complaining)

**Comment by ThatKoza** [comment]
> > > in fact, it did something for me. camera isnt getting recognised at all now, lol. you propably should make a seperate release for macos lower than mojave, because it just looks like it breaks in newer versions.
> > 
> > 
> > SORRY I made a typo in the script LOL. This one should work.
> > [SA_system_emulation.min.zip](https://github.com/user-attachments/files/18883824/SA_system_emulation.min.zip)
> 
> ok, now it works, ahahahahah. then i doubt that it needs a separate release (unless someone with old macos will be complaining)

also, it looks that tracking got a little better, even with my shit lighting, dunno if you did anything on your end with that

**Comment by ButzYung** [comment]
I don't know. Was there a popup asking for camera permission? Maybe when the permission is granted some camera features will become unrestricted.

**Comment by ThatKoza** [comment]
> I don't know. Was there a popup asking for camera permission? Maybe when the permission is granted some camera features will become unrestricted.

since i previously already gave electron camera permission, it wont ask for permission again, unless im doing a fresh install of the electron

**Comment by ButzYung** [comment]
@ThatKoza I have PMed you the first beta version of XR Animator for macOS. Please check your discord.
